sdk/core/azure-core/src/http/websockets/websockets_impl.cpp

```diff
@@ -0,0 +1,884 @@
+      m_initialBodyStream = response->ExtractBodyStream();
+      m_pingThread.Start(m_transport);
+    }
+
+    // Remember the protocol that the client chose.
+    auto chosenProtocol = responseHeaders.find("Sec-WebSocket-Protocol");
+    if (chosenProtocol != responseHeaders.end())
+    {
+      m_chosenProtocol = chosenProtocol->second;
+    }
+
+    m_state = SocketState::Open;
+  }
+  bool WebSocketImplementation::HasBuiltInWebSocketSupport()
+  {
+    std::lock_guard<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+    if (m_state != SocketState::Open)
+    {
+      throw std::runtime_error("Socket is not open.");
+    }
+    return m_transport->HasBuiltInWebSocketSupport();
+  }
+
+  std::string const& WebSocketImplementation::GetNegotiatedProtocol()
+  {
+    std::lock_guard<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+    if (m_state != SocketState::Open)
+    {
+      throw std::runtime_error("Socket is not open.");
+    }
+    return m_chosenProtocol;
+  }
+
+  void WebSocketImplementation::AddHeader(std::string const& header, std::string const& headerValue)
+  {
+    std::lock_guard<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+    if (m_state != SocketState::Closed && m_state != SocketState::Invalid)
+    {
+      throw std::runtime_error("AddHeader can only be called on closed sockets.");
+    }
+    m_headers.emplace(std::make_pair(header, headerValue));
+  }
+
+  void WebSocketImplementation::Close(
+      uint16_t closeStatus,
+      std::string const& closeReason,
+      Azure::Core::Context const& context)
+  {
+    std::unique_lock<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+
+    // If we're closing an already closed socket, we're done.
+    if (m_state == SocketState::Closed)
+    {
+      return;
+    }
+    if (m_state != SocketState::Open)
+    {
+      throw std::runtime_error("Socket is not open.");
+    }
+    m_state = SocketState::Closing;
+#if SUPPORT_NATIVE_TRANSPORT
+    if (m_transport->HasBuiltInWebSocketSupport())
+    {
+      m_transport->NativeCloseSocket(closeStatus, closeReason.c_str(), context);
+    }
+    else
+#endif
+    {
+      // Send a going away message to the server.
+      std::vector<uint8_t> closePayload;
+      closePayload.push_back(closeStatus >> 8);
+      closePayload.push_back(closeStatus & 0xff);
+      closePayload.insert(closePayload.end(), closeReason.begin(), closeReason.end());
+      std::vector<uint8_t> closeFrame = EncodeFrame(SocketOpcode::Close, true, closePayload);
+      SendTransportBuffer(closeFrame, context);
+
+      // Unlock the state mutex before waiting for the close response to be received.
+      lock.unlock();
+      // To ensure that we process the responses in a "timely" fashion, limit the close
+      // reception to 20 seconds if we don't already have a timeout.
+      Azure::Core::Context closeContext = context;
```

Comment:
Its possible the timeout could be something > than 20 seconds.  It seems reasonable to always put a 20s timeout on the request.

Reply:
The reason for this is that we have a test which uses Windows WebSockets to talk to http://www.microsoft.com and that takes a very long time to complete (about 2 minutes) because the HTTP transport keeps trying to retry the DNS error.

Reply:
Removing the timeout.